File: src/main/java/edu/kit/iti/formal/stvs/view/common/PositiveIntegerInputField.java
package edu.kit.iti.formal.stvs.view.common;

import edu.kit.iti.formal.stvs.view.ViewUtils;

import java.util.Optional;

import javafx.beans.property.BooleanProperty;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.beans.value.ObservableValue;
import javafx.geometry.Pos;
import javafx.scene.control.TextField;

/**
 * A Input field that only allows positive integers.
 *
 * @author Carsten Csiky
 */
public class PositiveIntegerInputField extends TextField {
  private BooleanProperty valid;

  /**
   * Creates an instances of this positive integer only field.
   */
  public PositiveIntegerInputField() {
    this.textProperty().addListener(this::onInputChange);
    valid = new SimpleBooleanProperty();
    valid.addListener(this::onValidStateChange);
    this.alignmentProperty().setValue(Pos.CENTER_RIGHT);
    ViewUtils.setupClass(this);
  }

  private void onValidStateChange(ObservableValue<?> observableValue, Boolean old, Boolean value) {
    if (value) {
      this.getStyleClass().add("valid");
    } else {
      this.getStyleClass().remove("valid");
    }
  }

  private void onInputChange(ObservableValue<?> observableValue, String old, String newValue) {
    valid.set(getText().trim().matches("(?!0)[0-9]+"));
  }

  /**
   * get inputfield value as an integer if no integer representation is available Optional.empty()
   * will be returned
   *
   * @return value as an integer
   */
  public Optional<Integer> getInteger() {
    if (valid.get()) {
      return Optional.of(Integer.valueOf(this.getText().trim()));
    } else {
      return Optional.empty();
    }
  }

  public boolean isValid() {
    return valid.get();
  }

  public BooleanProperty validProperty() {
    return valid;
  }
}

File: stvs/src/main/kotlin/edu/kit/iti/formal/stvs/view/common/PositiveIntegerInputField.kt
/* *****************************************************************
 * This file belongs to verifaps-lib (https://verifaps.github.io).
 * SPDX-License-Header: GPL-3.0-or-later
 * 
 * This program isType free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program isType distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a clone of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * *****************************************************************/
package edu.kit.iti.formal.stvs.view.common

import edu.kit.iti.formal.stvs.view.ViewUtils
import javafx.beans.property.BooleanProperty
import javafx.beans.property.SimpleBooleanProperty
import javafx.geometry.Pos
import javafx.scene.control.TextField

/**
 * A Input field that only allows positive integers.
 *
 * @author Carsten Csiky
 */
class PositiveIntegerInputField : TextField() {
    private val valid: BooleanProperty

    /**
     * Creates an instances of this positive integer only field.
     */
    init {
        textProperty().addListener { observableValue, old, newValue ->
            this.onInputChange(newValue)
        }
        valid = SimpleBooleanProperty()
        valid.addListener { _, _, value: Boolean ->
            this.onValidStateChange(value)
        }
        alignmentProperty().value = Pos.CENTER_RIGHT
        ViewUtils.setupClass(this)
    }

    private fun onValidStateChange(value: Boolean) {
        if (value) {
            styleClass.add("valid")
        } else {
            styleClass.remove("valid")
        }
    }

    private fun onInputChange(newValue: String) {
        valid.set(text.trim { it <= ' ' }.matches("(?!0)[0-9]+".toRegex()))
    }

    val integer: Int?
        /**
         * get inputfield value as an integer if no integer representation is available Optional.empty()
         * will be returned
         *
         * @return value as an integer
         */
        get() = if (valid.get()) {
            text.trim { it <= ' ' }.toInt()
        } else {
            null
        }

    fun isValid(): Boolean = valid.get()

    fun validProperty(): BooleanProperty = valid
}